src/v/ssx/abort_source.h

```diff
@@ -20,61 +21,106 @@
-                    [dex](auto& as) { return as.request_abort_ex(dex); });
-              });
-            if (sub) {
-                _sub.emplace(std::move(*sub));
-            }
-        });
-    }
-
-    ss::future<> stop() noexcept {
-        _sub.reset();
-        return request_abort().then([this]() { return _as.stop(); });
-    }
-
-    auto& local() noexcept { return _as.local(); }
-    auto const& local() const noexcept { return _as.local(); }
-
-    template<typename Func>
-    [[nodiscard]] auto subscribe(Func&& func) {
-        return local().subscribe(std::forward<Func>(func));
+    sharded_abort_source()
+      : _internal(std::make_shared<sharded_abort_source_internal>(
+        ss::this_shard_id(),
+        std::vector<ss::abort_source>(ss::smp::count),
+        std::nullopt)) {}
+
+    // Returns a reference to an abort_source local to the calling shard
+    //
+    // Note: the lifetime of `sharded_abort_source` must exceed the lifetime of
+    // this reference.
+    auto& local() noexcept { return _internal->as[ss::this_shard_id()]; }
+
+    // Returns a const reference to an abort_source local to the calling shard
+    //
+    // Note: the lifetime of `sharded_abort_source` must exceed the lifetime of
+    // this reference.
+    auto const& local() const noexcept {
+        return _internal->as[ss::this_shard_id()];
     }
 
+    // Requests an abort with exception_ptr `ex` on every shard's abort_source.
     ss::future<> request_abort_ex(std::exception_ptr ex) noexcept {
-        return _as.invoke_on_all(
-          [ex](auto& s) { return s.request_abort_ex(ex); });
+        return ss::smp::invoke_on_all(
+          [ex, as = *this]() mutable { as.local().request_abort_ex(ex); });
     }
 
+    // Requests an abort with exception `e` on every shard's abort_source.
     template<typename Exception>
     ss::future<> request_abort_ex(Exception&& e) noexcept {
         return request_abort_ex(std::make_exception_ptr(e));
     }
 
+    // Requests an abort with the default_exception type on every shard's
+    // abort_source.
     ss::future<> request_abort() noexcept {
         return request_abort_ex(local().get_default_exception());
     }
 
+    // Subscribes to the calling shard's local abort_source.
+    template<typename Func>
+    [[nodiscard]] auto subscribe(Func&& func) {
+        return local().subscribe(std::forward<Func>(func));
+    }
+
+    // Checks the calling shard's abort_source to see if an abort was requested.
     auto abort_requested() const noexcept { return local().abort_requested(); }
+
+    // Throws the abort exception of the calling shard's abort_source if its
+    // been aborted.
     auto check() const { return local().check(); }
 
-    bool local_is_initialized() const noexcept {
-        return _as.local_is_initialized();
+    // Removes subscription to parent abort source and requests an abort on
+    // every shard's abort_source.
+    ss::future<> stop() noexcept {
```

Comment:
It isn't clear to me the purpose of this method any more.

It resets the smart pointer and then calls `request_abort()`. So why not just call `request_abort()`? The smart pointer will be "reset" as soon as the object is destroyed any way. So `stop` doesn't seem to provide much functionality.